source-container-build/app/source_build.py

```diff
@@ -528,6 +438,317 @@ def parse_image_name(image: str) -> tuple[str, str, str]:
+        if cur_checksum == checksum:
+            return self
+
+        origin_d = self.descriptor
+        new_d: DescriptorT = {
+            "mediaType": origin_d["mediaType"],
+            "digest": f"sha256:{checksum}",
+            "size": 0,  # will be set later
+        }
+        if "annotations" in origin_d:
+            new_d["annotations"] = origin_d["annotations"]
+
+        new_blob = self.__class__(self._layout, new_d)
+        new_blob.path.write_bytes(self._raw_content)
+        new_blob.descriptor["size"] = new_blob.path.stat().st_size
+        return new_blob
+
+
+class Layer(Blob):
+    """Represent an image layer. Currently, no operation on the layers."""
+
+
+class JSONBlob(Blob):
+    """A blob whose content is encoded in JSON"""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self._python_obj: dict | None = None
+
+    def save(self) -> Blob:
+        """Write JSON string in text mode"""
+        self.raw_content = json.dumps(self.to_python).encode("utf-8")
+        return super().save()
+
+    @property
+    def to_python(self) -> dict[str, Any]:
+        if self._python_obj is None:
+            self._python_obj = json.loads(super().raw_content)
+        return self._python_obj
+
+
+class Config(JSONBlob):
+    """Image config"""
+
+    @property
+    def history(self) -> list[HistoryT]:
+        """Return .history"""
+        return self.to_python["history"]
+
+    @property
+    def diff_ids(self) -> list[str]:
+        """Return .rootfs.diff_ids"""
+        return self.to_python["rootfs"]["diff_ids"]
+
+
+class Manifest(JSONBlob):
+    """Image manifest"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._config: Config | None = None
+        self._layers: list[Layer] | None = None
+
+    @property
+    def config(self) -> Config:
+        if self._config is None:
+            self._config = Config(self._layout, self.to_python["config"])
+        return self._config
+
+    @property
+    def layers(self) -> list[Layer]:
+        if self._layers is None:
+            self._layers = [Layer(self._layout, d) for d in self.to_python["layers"]]
+        return self._layers
+
+    def append_layer(self, layer: Layer) -> None:
+        layers: list[DescriptorT] = self.to_python["layers"]
+        layers.append(layer.descriptor)
+
+    def save(self) -> Blob:
+        """Save this manifest"""
+        new_config = self.config.save()
+        if new_config != self.config:
+            self.to_python["config"] = new_config.descriptor
+        layer_descriptors = self.to_python["layers"]
+        for idx, layer in enumerate(self.layers):
+            if not layer.path.exists():
+                raise ValueError(f"layer {str(layer.path)} does not exist.")
+            new_layer = layer.save()
+            if new_layer != layer:
+                layer_descriptors[idx] = new_layer.descriptor
+                layer.delete()
```

Comment:
> This is intended. The modification to the layers will be available in the new manifest after calling save. The basic idea is, changes to a blob will be available in the new blob after save, and the current blob always references the current one.

In that case, I probably don't understand why this is needed. If `self.layers` references the original layers, is this chunk of code needed? And now that we're prepending the parent layers instead of appending, does this work correctly?

Reply:
Good catch. Fixed.